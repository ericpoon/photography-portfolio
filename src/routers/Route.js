import React from 'react';
import { Route as NativeRoute } from 'react-router-dom';

const Route = (props) => {
  const { component: Component, header: Header, ...restProps } = props;
  const RouteComponent = (componentProps) => {
    if (Header) {
      return (
        <div>
          <Header />
          <Component {...componentProps} />
        </div>
      );
    }
    return <Component {...componentProps} />;
  };

  return <NativeRoute {...restProps} component={RouteComponent} />;
};

export default Route;
