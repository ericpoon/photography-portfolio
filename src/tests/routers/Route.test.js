import React from 'react';
import { shallow } from 'enzyme';
import Route from '../../routers/Route';

let wrapper,
  Header,
  Component,
  props;

beforeEach(() => {
  Header = jest.fn(() => (<div className={'header'} />));
  Component = jest.fn(() => (<div className={'component'} />));
  props = { prop1: 'a prop', prop2: 'another prop' };
  wrapper = shallow(
    <Route
      header={Header}
      component={Component}
      {...props}
    />);
});

describe('with Header provided', () => {
  it('renders Route', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should have Header in Route', () => {
    const routeComponentWrapper = shallow(wrapper.prop('component')());
    expect(
      routeComponentWrapper
        .children()
        .map(c => c.dive())
        .filter(c => c.equals(Header())))
      .toHaveLength(1);
  });

  it('should have Component in Route', () => {
    const routeComponentWrapper = shallow(wrapper.prop('component')());
    expect(
      routeComponentWrapper
        .children()
        .map(c => c.dive())
        .filter(c => c.equals(Component())))
      .toHaveLength(1);
  });

  it('should pass props to native Route from react-router-dom', () => {
    expect(wrapper.props()).toMatchObject(props);
  });
});

describe('without Header provided', () => {
  beforeEach(() => {
    wrapper.setProps({ header: null });
  });

  it('should not have Header in Route', () => {
    const routeComponentWrapper = shallow(wrapper.prop('component')());
    expect(
      routeComponentWrapper
        .children()
        .map(c => c.dive())
        .filter(c => c.equals(Header())))
      .toHaveLength(0);
  });

  it('should have Component in Route', () => {
    const routeComponentWrapper = shallow(wrapper.prop('component')());
    expect(routeComponentWrapper.equals(Component())).toBeTruthy();
  });

  it('should pass props to native Route from react-router-dom', () => {
    expect(wrapper.props()).toMatchObject(props);
  });
});

