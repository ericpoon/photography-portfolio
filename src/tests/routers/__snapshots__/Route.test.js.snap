// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`with Header provided renders Route 1`] = `
<Route
  component={[Function]}
  prop1="a prop"
  prop2="another prop"
/>
`;
